lib/bigdecimal/math.rb

```diff
@@ -131,6 +132,26 @@ def cos(x, prec)
     y
   end
 
+  # call-seq:
+  #   tan(decimal, numeric) -> BigDecimal
+  #
+  # Computes the tangent of +decimal+ to the specified number of digits of
+  # precision, +numeric+.
+  #
+  # If +decimal+ is Infinity or NaN, returns NaN.
+  #
+  #   BigMath.tan(BigDecimal("0.0"), 4).to_s
+  #   #=> "0.0"
+  #
+  #   BigMath.tan(BigMath.PI(4) / 4, 4).to_s
+  #   #=> "0.999999999999999999999955588155008544487055622030633191403625599381672572e0"
+  #
+  def tan(x, prec)
+    sine = sin(x, prec)
+    cosine = sqrt(1 - sine**2, prec)
+    sine / cosine
```

Comment:
As I wrote in #319, I want to redefine `sin` by $\sin x = \sqrt{1 - \cos^2 x}$ for $|x| \le \pi$.
So I want to rewrite these lines as follow:

```rb
    x, sign = adjust_x_and_detect_sgin_of_tangent(x)
    t = guarantee_precision(prec) do |n|
      c = cos(x, n)
      s = sqrt(1 - c**2, n)
      s.div(c, n)
    end
    sign >= 0 ? t : -t
```

The `adjust_x_and_detect_sign_of_tangent` method adjusts $x$ within the range $[-\pi, \pi]$ and calculates the result sign from the original $x$.
The `guarantee_precision` method performs Ziv's loop to ensure the result of the block has the desired precision.